File: src/lib/api/profitAnalysis.ts
import { apiGet } from '@/lib/api/client.ts'

export type ProfitMetric = {
  value: number
  previous_value: number
  change_amount: number
  change_percent: number | null
}

export type ProfitTrendRow = {
  month: string
  channel_settlement: number
  rd_cost: number
  server_cost: number
  operating_expense: number
  operating_profit: number
  profit_margin: number
}

export type ProfitExpenseCategoryRow = {
  category: string
  amount: number
  share_percent: number
}

export type ProfitGameRow = {
  game_name: string
  channel_settlement: number
  rd_cost: number
  server_cost_allocated: number
  attributed_expense: number
  attributable_profit: number
  attributable_margin: number
  channel_flow: number
  rd_flow: number
}

export type ProfitAnalysis = {
  month: string
  previous_month: string
  available_months: string[]
  latest_month?: string | null
  channel_settlement: ProfitMetric
  rd_cost: ProfitMetric
  server_cost: ProfitMetric
  legacy_server_cost: ProfitMetric
  standalone_server_cost: ProfitMetric
  shared_server_cost: ProfitMetric
  attributed_server_cost: ProfitMetric
  operating_expense: ProfitMetric
  pre_expense_contribution: ProfitMetric
  operating_profit: ProfitMetric
  profit_margin: ProfitMetric
  shared_expense: ProfitMetric
  attributed_expense: ProfitMetric
  channel_bill_count: number
  rd_bill_count: number
  server_cost_count: number
  expense_count: number
  expense_categories: ProfitExpenseCategoryRow[]
  games: ProfitGameRow[]
  trend: ProfitTrendRow[]
  notes: string[]
}

export type ProjectProfitMonthRow = {
  month: string
  channel_settlement: number
  rd_cost: number
  server_cost: number
  attributed_expense: number
  total_attributable_cost: number
  gross_profit: number
  gross_margin: number
}

export type ProjectProfitRow = {
  game_name: string
  channel_settlement: number
  rd_cost: number
  server_cost: number
  attributed_expense: number
  total_attributable_cost: number
  gross_profit: number
  gross_margin: number
  channel_flow: number
  rd_flow: number
  active_months: number
  first_month: string | null
  last_month: string | null
  monthly: ProjectProfitMonthRow[]
}

export type ProjectProfitAnalysis = {
  scope: 'year' | 'lifetime'
  year: string | null
  available_years: string[]
  summary: {
    project_count: number
    profitable_projects: number
    loss_projects: number
    channel_settlement: number
    total_attributable_cost: number
    gross_profit: number
    gross_margin: number
    shared_server_cost: number
    shared_expense: number
    data_months: number
  }
  projects: ProjectProfitRow[]
  notes: string[]
}

export function getProfitAnalysis(params: {
  month?: string
  trendMonths?: number
} = {}): Promise<ProfitAnalysis> {
  const query = new URLSearchParams()
  if (params.month) query.set('month', params.month)
  if (params.trendMonths != null) query.set('trend_months', String(params.trendMonths))
  const qs = query.toString()
  return apiGet(`/api/profit-analysis/monthly${qs ? `?${qs}` : ''}`)
}

export function getProjectProfitAnalysis(params: {
  year?: string | null
} = {}): Promise<ProjectProfitAnalysis> {
  const query = new URLSearchParams()
  if (params.year) query.set('year', params.year)
  const qs = query.toString()
  return apiGet(`/api/profit-analysis/projects${qs ? `?${qs}` : ''}`)
}

File: backend/app/schemas/profit_analysis.py
"""经营利润分析 API 模型。"""

from __future__ import annotations

from pydantic import BaseModel, Field


class ProfitMetric(BaseModel):
    value: float = 0
    previous_value: float = 0
    change_amount: float = 0
    change_percent: float | None = None


class ProfitTrendRow(BaseModel):
    month: str
    channel_settlement: float = 0
    rd_cost: float = 0
    server_cost: float = 0
    operating_expense: float = 0
    operating_profit: float = 0
    profit_margin: float = 0


class ProfitExpenseCategoryRow(BaseModel):
    category: str
    amount: float = 0
    share_percent: float = 0


class ProfitGameRow(BaseModel):
    game_name: str
    channel_settlement: float = 0
    rd_cost: float = 0
    server_cost_allocated: float = 0
    attributed_expense: float = 0
    attributable_profit: float = 0
    attributable_margin: float = 0
    channel_flow: float = 0
    rd_flow: float = 0


class ProfitAnalysisRead(BaseModel):
    month: str
    previous_month: str
    available_months: list[str] = Field(default_factory=list)
    latest_month: str | None = None

    channel_settlement: ProfitMetric
    rd_cost: ProfitMetric
    server_cost: ProfitMetric
    legacy_server_cost: ProfitMetric = Field(default_factory=ProfitMetric)
    standalone_server_cost: ProfitMetric = Field(default_factory=ProfitMetric)
    shared_server_cost: ProfitMetric = Field(default_factory=ProfitMetric)
    attributed_server_cost: ProfitMetric = Field(default_factory=ProfitMetric)
    operating_expense: ProfitMetric
    pre_expense_contribution: ProfitMetric
    operating_profit: ProfitMetric
    profit_margin: ProfitMetric
    shared_expense: ProfitMetric
    attributed_expense: ProfitMetric

    channel_bill_count: int = 0
    rd_bill_count: int = 0
    server_cost_count: int = 0
    expense_count: int = 0

    expense_categories: list[ProfitExpenseCategoryRow] = Field(default_factory=list)
    games: list[ProfitGameRow] = Field(default_factory=list)
    trend: list[ProfitTrendRow] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)
